### ISSUE TITLE
CastError: Cast to ObjectID failed with discriminators/document arrays

### ISSUE THREAD

**Issue by saveman71** [Automattic/mongoose#10729]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

The following script is a valid usage of document arrays with discriminators. It **works** with mongoose 5. In mongoose 6 it throws

```
CastError: Cast to ObjectID failed for value "{
  title: 'Product title', 
 _id: new ObjectId("6141f19742a66dd9e5b6de98")
}" (type model) at path "product" 
  at EmbeddedDocument.$set (/home/arch/git/repro/node_modules/mongoose/lib/document.js:1427:9)
  at EmbeddedDocument.$set (/home/arch/git/repro/node_modules/mongoose/lib/document.js:1122:16)
  at EmbeddedDocument.Document (/home/arch/git/repro/node_modules/mongoose/lib/document.js:150:12)
  at EmbeddedDocument.Subdocument (/home/arch/git/repro/node_modules/mongoose/lib/types/subdocument.js:28:12)
  at EmbeddedDocument.ArraySubdocument [as constructor] (/home/arch/git/repro/node_modules/mongoose/lib/types/ArraySubdocument.js:38:15)
  at new ProductItem (/home/arch/git/repro/node_modules/mongoose/lib/schema/documentarray.js:115:17) at Object.<anonymous> (/home/arch/git/repro/repro-mongoose.js:35:21) 
  at Module._compile (internal/modules/cjs/loader.js:1072:14) 
  at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10) 
  at Module.load (internal/modules/cjs/loader.js:937:32)
```

```js
// Modules
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const ProductSchema = new Schema({
  title: String,
});
const Product = mongoose.model('Product', ProductSchema);
const ProductItemSchema = new Schema({
  product: { type: Schema.Types.ObjectId, ref: 'Product' },
});

const OrderItemSchema = new mongoose.Schema({
  quantity: Number,
}, {discriminatorKey: '__t'});
const OrderItem = mongoose.model('OrderItem', OrderItemSchema);


const OrderSchema = new Schema({
  items: {
    type: [OrderItemSchema],
    default: [],
  },
});

const docArray = OrderSchema.path('items');

const ProductItem = docArray
  .discriminator('ProductItem', ProductItemSchema);

const Order = mongoose.model('Order', OrderSchema);

const product = new Product({title: 'Product title'});

const productItem = new ProductItem({
  product,
})
```

The underlying error is

```
TypeError: Cannot read property 'endsWith' of null
      at EmbeddedDocument.Document.populated (/home/arch/git/repro/node_modules/mongoose/lib/document.js:4135:24)
      at Function.SchemaType._isRef (/home/arch/git/repro/node_modules/mongoose/lib/schematype.js:1388:17)
      at ObjectId.cast (/home/arch/git/repro/node_modules/mongoose/lib/schema/objectid.js:225:18)
      at ObjectId.SchemaType.applySetters (/home/arch/git/repro/node_modules/mongoose/lib/schematype.js:1122:12)
      at EmbeddedDocument.$set (/home/arch/git/repro/node_modules/mongoose/lib/document.js:1380:20)
      at EmbeddedDocument.$set (/home/arch/git/repro/node_modules/mongoose/lib/document.js:1122:16)
      at EmbeddedDocument.Document (/home/arch/git/repro/node_modules/mongoose/lib/document.js:150:12)
      at EmbeddedDocument.Subdocument (/home/arch/git/repro/node_modules/mongoose/lib/types/subdocument.js:28:12)
      at EmbeddedDocument.ArraySubdocument [as constructor] (/home/arch/git/repro/node_modules/mongoose/lib/types/ArraySubdocument.js:38:15)
      at new ProductItem (/home/arch/git/repro/node_modules/mongoose/lib/schema/documentarray.js:115:17)
```

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

MongoDB 3.4
Mongoose 6.0.5
NodeJS v14.17.6

**Comment by saveman71** [comment]
Note that `{clone: false}` (the only mention of discriminators in the migration doc) doesn't change anything.

**Comment by IslandRhythms** [comment]
```javascript
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const ProductSchema = new Schema({
  title: String,
});
const Product = mongoose.model('Product', ProductSchema);
const ProductItemSchema = new Schema({
  product: { type: Schema.Types.ObjectId, ref: 'Product' },
});

const OrderItemSchema = new mongoose.Schema({
  quantity: Number,
}, {discriminatorKey: '__t'});
const OrderItem = mongoose.model('OrderItem', OrderItemSchema);


const OrderSchema = new Schema({
  items: {
    type: [OrderItemSchema],
    default: [],
  },
});

const docArray = OrderSchema.path('items');

async function run() {
    await mongoose.connect('mongodb://localhost:27017/', {useNewUrlParser: true,
    useUnifiedTopology: true,});
    await mongoose.connection.dropDatabase();

    const ProductItem = docArray
  .discriminator('ProductItem', ProductItemSchema);

const Order = mongoose.model('Order', OrderSchema);

const product = new Product({title: 'Product title'});

const productItem = new ProductItem({
  product,
})
}

run();
```

**Comment by vkarpov15** [comment]
Thanks for your patience, I can confirm this was fixed in v6.0.7 with #10749. Please upgrade to v6.0.7 and this issue should go away.
